monitor/monitor.go

```diff
@@ -119,172 +119,29 @@ func RunMonitor(ctx context.Context, curRef string, options *controllerapi.Build
-					if err == nil && isProcess {
-						if err := c.DisconnectProcess(ctx, curRef, target); err != nil {
-							fmt.Printf("disconnect process failed %v\n", target)
-							continue
-						}
-						continue
-					}
-					if err := c.Disconnect(ctx, target); err != nil {
-						fmt.Println("disconnect error", err)
-					}
-				case "kill":
-					if err := c.Kill(ctx); err != nil {
-						fmt.Printf("failed to kill: %v\n", err)
-					}
-				case "attach":
-					if len(args) < 2 {
-						fmt.Println("attach: server name must be passed")
-						continue
-					}
-					ref := args[1]
-					var id string
-
-					isProcess, err := isProcessID(ctx, c, curRef, ref)
-					if err == nil && isProcess {
-						m.attach(ctx, curRef, ref)
-						id = ref
-					}
-					if id == "" {
-						refs, err := c.List(ctx)
-						if err != nil {
-							fmt.Printf("failed to get the list of sessions: %v\n", err)
-							continue
-						}
-						found := false
-						for _, s := range refs {
-							if s == ref {
-								found = true
-								break
-							}
-						}
-						if !found {
-							fmt.Printf("unknown ID: %q\n", ref)
-							continue
-						}
-						if m.invokeCancel != nil {
-							m.invokeCancel() // Finish existing attach
-						}
-						curRef = ref
-					}
-					fmt.Fprintf(stdout, "Attached to process %q. Press Ctrl-a-c to switch to the new container\n", id)
-				case "exec":
-					if len(args) < 2 {
-						fmt.Println("attach: command must be passed")
-						continue
-					}
-					if curRef == "" {
-						fmt.Println("attach to a session first")
-						continue
-					}
-					cfg := controllerapi.InvokeConfig{
-						Entrypoint: []string{args[1]},
-						Cmd:        args[2:],
-						// TODO: support other options as well via flags
-						Env:  invokeConfig.Env,
-						User: invokeConfig.User,
-						Cwd:  invokeConfig.Cwd,
-						Tty:  true,
-					}
-					pid := m.exec(ctx, curRef, cfg)
-					fmt.Fprintf(stdout, "Process %q started. Press Ctrl-a-c to switch to that process.\n", pid)
-				case "ps":
-					plist, err := c.ListProcesses(ctx, curRef)
-					if err != nil {
-						fmt.Println("cannot list process:", err)
-						continue
-					}
-					tw := tabwriter.NewWriter(stdout, 1, 8, 1, '\t', 0)
-					fmt.Fprintln(tw, "PID\tCURRENT_SESSION\tCOMMAND")
-					for _, p := range plist {
-						fmt.Fprintf(tw, "%-20s\t%v\t%v\n", p.ProcessID, p.ProcessID == m.attachedPid.Load(), append(p.InvokeConfig.Entrypoint, p.InvokeConfig.Cmd...))
-					}
-					tw.Flush()
+					continue
 				case "exit":
 					return
 				case "help":
-					fmt.Fprint(stdout, helpMessage)
+					printHelpMessage(stdout, registeredCommands)
+					continue
 				default:
+				}
+
+				// Registered commands
+				cmdname := args[0]
+				if cm, ok := registeredCommands[cmdname]; ok {
+					if err := cm.Exec(ctx, args); err != nil {
+						fmt.Fprintf(stdout, "failed %q: %v\n", cmdname, err)
+					}
+				} else {
 					fmt.Printf("unknown command: %q\n", l)
```

Comment:
For errors that aren't from a specific command, could we prefix them with `monitor:`? That way it kind of mirrors `sh:` errors as opposed to `cat:` errors.